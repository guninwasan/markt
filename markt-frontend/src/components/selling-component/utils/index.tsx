export * from "./upload-image";
export * from "./valid-form-details";
