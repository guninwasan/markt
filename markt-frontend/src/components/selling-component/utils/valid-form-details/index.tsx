export * from "./valid-form-details";
