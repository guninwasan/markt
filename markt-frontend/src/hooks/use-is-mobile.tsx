import { useEffect, useState } from "react";

const MOBILE_BREAKPOINT = 770;

const useIsMobile = () => {
  const [isMobile, setIsMobile] = useState<null | boolean>(null);

  useEffect(() => {
    const handleResize = () => {
      setIsMobile(window.innerWidth <= MOBILE_BREAKPOINT);
    };

    window.addEventListener("resize", handleResize);

    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, []);

  return { isMobile };
};

export { useIsMobile };
