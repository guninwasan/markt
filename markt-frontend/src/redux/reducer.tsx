import { combineReducers } from "@reduxjs/toolkit";
import userAuth from "./slices/user-auth-slice";
import device from "./slices/device-slice";

const reducer = combineReducers({
  userAuth,
  device,
});

export { reducer };
