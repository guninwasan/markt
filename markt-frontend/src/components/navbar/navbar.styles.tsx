import styled from "styled-components";
import { NavLink } from "react-router-dom";
import { FaSearch } from "react-icons/fa";
import { colors } from "../../utils";

const NavbarContainer = styled.nav`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1.5rem 2.5rem;
  background-color: ${colors.primary};
  color: ${colors.textBlack};
`;

const MarktHeaderText = styled(NavLink)`
  font-size: 1.5rem;
  font-weight: 1000;
  color: ${colors.textBlack};
  text-transform: uppercase;
  letter-spacing: 2px;
  text-decoration: none;
`;

const Menu = styled.ul<{ isMobile?: boolean | null }>`
  list-style: none;
  display: flex;
  margin: 0;
  padding: 0;
  gap: ${({ isMobile }) => (isMobile ? "2rem" : "0")};
  flex-direction: ${({ isMobile }) => (isMobile ? "column" : "row")};
`;

const MenuItem = styled.li`
  margin-left: 1rem;
  cursor: pointer;
  position: relative;
`;

const MenuLink = styled(NavLink)`
  color: ${colors.textBlack};
  text-decoration: none;
  padding: 0.5rem 1rem;

  &.active {
    background-color: #555;
    border-radius: 4px;
    color: #f9f9f9;
  }
`;

const LogoutButton = styled.div`
  color: ${colors.textBlack};
  text-decoration: none;
  padding: 0.5rem 1rem;
`;

const Hamburger = styled.div<{ isOpen: boolean }>`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  width: 1.5rem;
  height: 1.5rem;
  cursor: pointer;

  span {
    width: 1.5rem;
    height: 0.2rem;
    background: ${colors.black};
    border-radius: 10px;
    transition: all 0.3s linear;
    position: relative;
    transform-origin: 1px;
  }

  span:nth-child(1) {
    transform: ${({ isOpen }) => (isOpen ? "rotate(45deg)" : "rotate(0)")};
  }

  span:nth-child(2) {
    opacity: ${({ isOpen }) => (isOpen ? "0" : "1")};
    transform: ${({ isOpen }) =>
      isOpen ? "translateX(20px)" : "translateX(0)"};
  }

  span:nth-child(3) {
    transform: ${({ isOpen }) => (isOpen ? "rotate(-45deg)" : "rotate(0)")};
  }
`;

const MobileMenuContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Backdrop = styled.div<{ isOpen: boolean }>`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: ${({ isOpen }) => (isOpen ? "block" : "none")};
  z-index: 999;
`;

const Sidebar = styled.div<{ isOpen: boolean }>`
  height: 100vh;
  width: 90vw;
  background-color: ${colors.lightGrey};
  position: fixed;
  top: 4.5rem;
  right: 0;
  transform: ${({ isOpen }) => (isOpen ? "translateX(0)" : "translateX(100%)")};
  transition: transform 0.3s ease-in-out;
  z-index: 1000;
  padding: 2rem;
  display: flex;
  flex-direction: column;
`;

const ProfileContainer = styled.div`
  display: flex;
  align-items: center;
  gap: 0.25rem;
  flex-direction: column;
`;

const ProfileButton = styled.button`
  background-color: #e0e0e0;
  border: none;
  cursor: pointer;
  padding: 0;
  position: relative;
  border-radius: 50%;
  width: 40px;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;

  &:hover {
    background-color: #d0d0d0;
  }

  svg {
    width: 24px;
    height: 24px;
    fill: #333;
  }
`;

const ProfileIconText = styled.span`
  font-size: 0.75rem;
  color: ${colors.textBlack};
`;

const Dropdown = styled.div`
  position: absolute;
  top: 100%;
  right: 0;
  transform: translateX(calc(0%));
  background-color: ${colors.white};
  border: 1px solid #ddd;
  border-radius: 4px;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  z-index: 1001;
  padding: 0.5rem 0;
  width: 150px;
`;

const DropdownItem = styled.a`
  display: block;
  padding: 0.5rem 1rem;
  color: ${colors.textBlack};
  text-decoration: none;

  &:hover {
    background-color: #f0f0f0;
    color: ${colors.textBlack};
  }
`;

const SearchWithMenu = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
`;

const SearchButtonMobile = styled(FaSearch)`
  width: 1.25rem;
  height: 1.25rem;
  cursor: pointer;
`;

const ModalBackdrop = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1000;
`;

const SpotlightModal = styled.div`
  background-color: #f0f0f0;
  width: 100%;
  max-width: 300px;
  padding: 1rem;
  border-radius: 8px;
  box-shadow: 0 4px 10px rgba(0, 0, 0, 0.2);
  z-index: 1001;
  min-height: 100px;
`;

export {
  NavbarContainer,
  MarktHeaderText,
  Menu,
  MenuItem,
  MenuLink,
  Hamburger,
  MobileMenuContainer,
  Backdrop,
  Sidebar,
  ProfileButton,
  ProfileIconText,
  Dropdown,
  DropdownItem,
  ProfileContainer,
  LogoutButton,
  SearchWithMenu,
  SearchButtonMobile,
  ModalBackdrop,
  SpotlightModal,
};
