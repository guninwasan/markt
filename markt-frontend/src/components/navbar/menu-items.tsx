import { useState } from "react";
import { useIsMobile } from "../../hooks";
import {
  Menu,
  MenuItem,
  MenuLink,
  ProfileButton,
  Dropdown,
  DropdownItem,
} from "./navbar.styles";
import { NavLink } from "react-router-dom";

const MenuItems = () => {
  const { isMobile } = useIsMobile();
  const [isLoggedIn, setIsLoggedIn] = useState(true);
  const [showDropdown, setShowDropdown] = useState(false);

  const toggleDropdown = () => {
    setShowDropdown(!showDropdown);
  };

  const handleLogout = () => {
    setIsLoggedIn(false);
  };

  return (
    <Menu isMobile={isMobile}>
      <MenuItem>
        <MenuLink
          as={NavLink}
          to="/sell"
          className={({ isActive }) => (isActive ? "active" : "")}
        >
          Sell
        </MenuLink>
      </MenuItem>
      {isLoggedIn ? (
        isMobile ? (
          <>
            <MenuItem as={NavLink} to="/account">
              My Account
            </MenuItem>
            <MenuItem as={NavLink} to="/wishlist">
              Wishlist
            </MenuItem>
            <MenuItem onClick={handleLogout}>Logout</MenuItem>
          </>
        ) : (
          <MenuItem>
            <ProfileButton
              onClick={toggleDropdown}
              aria-expanded={showDropdown}
            >
              Profile
            </ProfileButton>
            {showDropdown && (
              <Dropdown>
                <DropdownItem as={NavLink} to="/account">
                  My Account
                </DropdownItem>
                <DropdownItem as={NavLink} to="/wishlist">
                  Wishlist
                </DropdownItem>
                <DropdownItem onClick={handleLogout}>Logout</DropdownItem>
              </Dropdown>
            )}
          </MenuItem>
        )
      ) : (
        <MenuItem>
          <MenuLink as={NavLink} to="/register">
            Login/Register
          </MenuLink>
        </MenuItem>
      )}
    </Menu>
  );
};

export { MenuItems };
