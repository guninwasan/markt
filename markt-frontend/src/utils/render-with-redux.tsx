import { render } from "@testing-library/react";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../redux";

export const renderWithRedux = (component: React.ReactNode) => {
  return render(
    <Provider store={store}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
};
