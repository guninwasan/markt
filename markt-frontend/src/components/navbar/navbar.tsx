import React, { useState } from "react";
import { useIsMobile } from "../../hooks";
import {
  NavbarContainer,
  MarktHeaderText,
  Hamburger,
  Backdrop,
  Sidebar,
} from "./navbar.styles";
import { MenuItems } from "./menu-items";

const Navbar = () => {
  const { isMobile } = useIsMobile();
  const [isOpen, setIsOpen] = useState(false);

  const toggleSidebar = () => {
    setIsOpen(!isOpen);
  };

  const closeSidebar = () => {
    setIsOpen(false);
  };

  return (
    <NavbarContainer>
      <MarktHeaderText to={"/"}>Markt</MarktHeaderText>
      {isMobile ? (
        <>
          <Hamburger
            onClick={toggleSidebar}
            isOpen={isOpen}
            data-testid="hamburger"
          >
            <span />
            <span />
            <span />
          </Hamburger>
          <Backdrop
            data-testid="backdrop"
            isOpen={isOpen}
            onClick={closeSidebar}
          />
          {isOpen && (
            <Sidebar data-testid="sidebar" isOpen={isOpen}>
              <MenuItems />
            </Sidebar>
          )}
        </>
      ) : (
        <MenuItems />
      )}
    </NavbarContainer>
  );
};

export { Navbar };
