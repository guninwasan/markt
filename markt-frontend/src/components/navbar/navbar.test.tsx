import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom"; // Import MemoryRouter
import { Navbar } from "./navbar";
import { store } from "../../redux/store";
import { useIsMobile } from "../../hooks";
import { renderWithRedux } from "../../utils/render-with-redux";

jest.mock("../../hooks", () => ({
  useIsMobile: jest.fn(),
}));

describe("Navbar", () => {
  const mockUseIsMobile = useIsMobile as jest.Mock;

  beforeEach(() => {
    mockUseIsMobile.mockReturnValue({ isMobile: false });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render the navbar properly", () => {
    renderWithRedux(<Navbar />);
    expect(document.querySelector("nav")).toBeInTheDocument();
  });

  it("should render the Markt name", () => {
    renderWithRedux(<Navbar />);
    const marktText = screen.getByText(/Markt/i);
    expect(marktText).toBeInTheDocument();
  });

  it("should render the Home, Buy, Sell link", () => {
    renderWithRedux(<Navbar />);
    const homeLink = screen.getByText(/Home/i);
    const buyLink = screen.getByText(/Buy/i);
    const sellLink = screen.getByText(/Sell/i);
    expect(homeLink).toBeInTheDocument();
    expect(buyLink).toBeInTheDocument();
    expect(sellLink).toBeInTheDocument();
  });
});
