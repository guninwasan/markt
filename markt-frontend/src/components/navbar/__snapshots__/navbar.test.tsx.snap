// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should match the snapshot 1`] = `
<nav
  class="sc-blHHSb nIHai"
>
  <div
    class="sc-gtLWhw dgXBwO"
  >
    Markt
  </div>
  <ul
    class="sc-egkSDF ckDOmE"
  >
    <li
      class="sc-fAUdSK kycBV"
    >
      <a
        class="sc-dntaoT lagFYI active"
        href="/"
      >
        Home
      </a>
    </li>
    <li
      class="sc-fAUdSK kycBV"
    >
      <a
        class="sc-dntaoT lagFYI"
        href="/"
      >
        Buy
      </a>
    </li>
    <li
      class="sc-fAUdSK kycBV"
    >
      <a
        class="sc-dntaoT lagFYI"
        href="/"
      >
        Sell
      </a>
    </li>
    <li
      class="sc-fAUdSK kycBV"
    >
      <a
        class="sc-dntaoT lagFYI"
        href="/"
      >
        Login/Register
      </a>
    </li>
  </ul>
</nav>
`;
